pyvirtualcam/native_macos/main.mm

```diff
@@ -0,0 +1,165 @@
+
+#include <stdexcept>
+#include <pybind11/pybind11.h>
+#include <pybind11/numpy.h>
+#include "OBSDALMachServer.h"
+#include "Defines.h"
+#include <mach/mach_time.h>
+#include <cstdint>
+
+namespace py = pybind11;
+
+#define UNUSED_PARAMETER(arg)
+
+
+static OBSDALMachServer *sMachServer = nil;
+
+static int cam_output_width;
+static int cam_output_height;
+static uint32_t cam_fps_num;
+static uint32_t cam_fps_den;
+
+
+static void YfromRGB(uint8_t* y, uint8_t r, uint8_t g, uint8_t b) {
+    *y = (uint8_t)( 0.257 * r + 0.504 * g + 0.098 * b +  16);
+}
+static void UVfromRGB(uint8_t* u, uint8_t* v, uint8_t r, uint8_t g, uint8_t b) {
+    *u = (uint8_t)(-0.148 * r - 0.291 * g + 0.439 * b + 128);
+    *v = (uint8_t)( 0.439 * r - 0.368 * g - 0.071 * b + 128);
+}
+
+static bool virtual_output_start(int width, int height, double fps, int delay) {
+    if (sMachServer != nil) {
+        fprintf(stderr, "virtual camera output already started\n");
+        return false;
+    }
+
+    cam_output_width = width;
+    cam_output_height = height;
+    cam_fps_num = fps * 1000;
+    cam_fps_den = 1000;
+
+	  blog(LOG_DEBUG, "output_create");
+	  sMachServer = [[OBSDALMachServer alloc] init];
+
+	  [sMachServer run];
+
+	  return true;
+}
+
+static void virtual_output_stop() {
+    if (sMachServer == nil) {
+        return;
+    }
+
+	  [sMachServer stop];
+
+	  sMachServer = nil;
+}
+
+// data is in RGBA format (packed RGBA 8:8:8:8, 32bpp, RGBARGBA...)
+// TODO RGB24 would be better but not supported in obs-virtual-cam
+static void virtual_output_raw_video(void *data, int linesize, uint64_t timestamp) {
+    if (sMachServer == nil) {
+        return;
+    }
+
+	  uint8_t *outData = (uint8_t *)data;
+	  if (linesize != (cam_output_width * 2)) {
+		    blog(LOG_ERROR,
+		         "unexpected frame->linesize (expected:%d actual:%d)",
+		         (int)(cam_output_width * 2), (int)linesize);
+        return;
+	  }
+
+	  CGFloat width = cam_output_width;
+	  CGFloat height = cam_output_height;
+
+	  [sMachServer sendFrameWithSize:NSMakeSize(width, height)
+			       timestamp:timestamp
+			    fpsNumerator:cam_fps_num
+			  fpsDenominator:cam_fps_den
+			      frameBytes:outData];
+}
+
+
+void start(int width, int height, double fps, int delay) {
+    if (!virtual_output_start(width, height, fps, delay))
+        throw std::runtime_error("error starting virtual camera output");
+}
+
+void send(py::array_t<uint8_t, py::array::c_style> frame) {
+    py::buffer_info buf = frame.request();
+    if (buf.ndim != 3)
+        throw std::runtime_error("ndim must be 3 (h,w,c)");
+    if (buf.shape[2] != 4)
+        throw std::runtime_error("frame must have 4 channels (rgba)");
+
+    // We must handle port messages, and somehow our RunLoop isn't normally active
+    NSRunLoop *runLoop;
+  	NSDate *timeout = [NSDate dateWithTimeIntervalSinceNow:0.001];
+    runLoop = [NSRunLoop currentRunLoop];
+    [runLoop runUntilDate:timeout];
```

Comment:
I wasted about 1-2 hours to find this solution.
I noticed that someone else had the exact same problem before (https://stackoverflow.com/questions/62253531/cannot-connect-to-nsmachbootstrapserver-from-plugin); at least I found a solution, but it feels very hacky.

There's probably a cleaner way - but I don't usually code on macOS and never really had to do ObjC before.

The timeout value (0.001) seems very janky, too.

Reply:
I assume the timeout is one of the reasons for failing CI performance tests. Not sure about the actual time-slice resolution, but I could imagine that this blocks for more than 1ms.

I did not check if it still works when providing  timeout in the past (so it returns immediately after handling messages).

Reply:
(Possibly) fixed in https://github.com/letmaik/pyvirtualcam/pull/22.